docs/specs/webauthn.md

```diff
@@ -0,0 +1,346 @@
+1. Authgear tells the platform all the passkeys the user already has, to prevent creating duplicate passkey.
+1. The user ends up with a new passkey.
+
+## Design decisions
+
+### WebAuthn credential is both an Identity and a Primary Authenticator
+
+The first reason is that if we ever need to support a use case of NOT showing the Login ID field for AutoFill,
+we can still look up a user when given a passkey because a passkey is an Authgear Identity.
+
+The second reason is to path the way for future support of using a passkey as a new 2FA factor.
+When a passkey is used as a 2FA factor, it is a WebAuthn Secondary Authenticator.
+Basically it is the same as WebAuthn Primary Authenticator except that its kind is secondary, thus
+can only be used in secondary authentication.
+
+### Passkey is unsupported by Authgear on platform without discoverable credentials
+
+Due to the decision that we treat every passkey as both Authgear Identity and Authgear Authenticator,
+we DO NOT support passkey on platform without [discoverable credentials](https://www.w3.org/TR/webauthn-2/#discoverable-credential).
+This simplifies our implementation because we can assume every passkey is discoverable.
+
+The notable platform that does not support discoverable credentials is [Chrome on Android](#authenticatorattachment-behavior-on-various-platforms).
+
+### The user either set up a password or a passkey during sign-up
+
+This signifies a passkey is a replacement of a password.
+Instead of always requiring the user to set up a password,
+the user can just set up a passkey only.
+This improves security by eliminating a weaker credential from the account.
+
+### The user can add a password by going through the reset password flow
+
+This helps the user who is not using Safari 16 to benefit from using passkeys,
+without the downside of permanent loss access to their account.
+
+## Future works
+
+### Allow the user to remove their password
+
+If the user has at least one passkey in their account, the user can remove their password.
+
+### Ensure passkey support for syncing cross-platform authenticator
+
+1Password has expressed their intended support for passkey in their blogpost https://blog.1password.com/1password-is-joining-the-fido-alliance/
+As of 2022-07-20, the build 80800215 of 1Password 8 still has no support for passkey that I can test with.
+From the video in the blogpost, the authenticator provided by 1Password is a cross-platform authenticator.
+It seems that 1Password will take care of syncing the passkeys across devices in their own way, just as iCloud Keychain does.
+
+We should keep an eye of the progress of 1Password passkey support to ensure it works seamlessly with Authgear.
+
+## Configuration
+
+The following configuration are added.
+
+```yaml
+authentication:
+    identities:
+    - login_id
+    - oauth
+    - webauthn
```

Comment:
I thought `webauthn` identity should be removed after update, so the authentication will become `login_id` + `webauthn`? 

But if `webauthn` identity is removed, what will happen if user add one more `login_id` in the setting. Which `login_id` will be used if the user authenticates with `webauthn` key?

Would like to know your idea about the model.

Reply:
> I thought webauthn identity should be removed after update, so the authentication will become login_id + webauthn?

A passkey is still considered as an identity and a primary authenticator. The reason is mentioned in "WebAuthn credential is both an Identity and a Primary Authenticator"

> But if webauthn identity is removed, what will happen if user add one more login_id in the setting. Which login_id will be used if the user authenticates with webauthn key?

We treat passkey as identity so this question does not apply.